src/arima.cpp

```diff
@@ -1,156 +1,212 @@
 #include <algorithm>
 #include <array>
+#include <cassert>
 #include <cmath>
+#include <span>
 #include <vector>
 
 #include <pybind11/numpy.h>
 #include <pybind11/pybind11.h>
 
 namespace arima {
 namespace py = pybind11;
+namespace {
+template <typename T> std::span<T> make_span(py::array_t<T> &array) {
+  // ssize_t to size_t is safe because the size of an array is non-negative
+  return {array.mutable_data(), static_cast<size_t>(array.size())};
+}
+
+template <typename T>
+std::span<const T> make_cspan(const py::array_t<T> &array) {
+  // ssize_t to size_t is safe because the size of an array is non-negative
+  return {array.data(), static_cast<size_t>(array.size())};
+}
+} // namespace
+
+void partrans(const uint32_t p, const std::span<const double> rawv,
+              const std::span<double> newv) {
+  assert(p <= rawv.size());
+  assert(p <= newv.size());
 
-void partrans(int p, const double *raw, double *newv) {
-  std::transform(raw, raw + p, newv, [](double x) { return std::tanh(x); });
-  std::vector<double> work(newv, newv + p);
-  for (int j = 1; j < p; ++j) {
-    for (int k = 0; k < j; ++k) {
+  std::transform(rawv.begin(), rawv.begin() + p, newv.begin(),
+                 [](double x) { return std::tanh(x); });
+
+  std::vector<double> work(newv.begin(), newv.begin() + p);
+  for (size_t j = 1; j < p; ++j) {
+    for (size_t k = 0; k < j; ++k) {
       work[k] -= newv[j] * newv[j - k - 1];
     }
-    std::copy(work.begin(), work.begin() + j, newv);
+    std::copy(work.begin(), work.begin() + j, newv.begin());
   }
 }
 
 std::tuple<py::array_t<double>, py::array_t<double>>
 arima_transpar(const py::array_t<double> params_inv,
-               const py::array_t<int> armav, bool trans) {
-  auto arma = armav.data();
-  auto params_in = params_inv.data();
-  int mp = arma[0];
-  int mq = arma[1];
-  int msp = arma[2];
-  int msq = arma[3];
-  int ns = arma[4];
-  int p = mp + ns * msp;
-  int q = mq + ns * msq;
-  auto params = std::vector<double>(params_in, params_in + params_inv.size());
+               const py::array_t<uint32_t> armav, bool trans) {
```

Comment:
Changed to uint32, did it together with the ssize_t fix :)

Reply:
Didn't expect that. I've changed it back to int32, it should be fine since they're all checked to be positive in python code, still added some asserts. I've kept the interfaces in `partrans` and `invpartrans`, since its just a trivial type and doesn't cause any issues like copies with np array.

It dawned on me that it could be a good idea to add checks for more uint/int overflows. But that shouldn't really happen unless you define models with insane orders/seasonality.